### ISSUE TITLE
Refund Box Grayed Out - Refunds Via Dashboard No Longer Possible

### ISSUE THREAD

**Issue by co50** [woocommerce/woocommerce#23434]
**Describe the bug**
After latest WC update, all orders normally eligible for in-dash refund have the input box for entering the refund amount grayed out. The same result occurs whether the order status in on processing, cancelled, on hold, or any other status. Before posting this, I quickly verified that multiple others users with 3rd party gateways such as Stripe, PayPal, etc. (which may or may not have anything to do with it) experience this same issue

**To Reproduce**
1. Go to 'Any new order (Status can be Processing, On Hold, or Cancelled)'
2. Click on 'REFUND'
3. Attempt to enter amount to be refunded

**Screenshots**
https://imgur.com/a/YlKqbth

**Expected behavior**
Amount to Refund box should not be grayed out and should allow input. 

**Isolating the problem (mark completed items with an [x]):**
- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
<details>
WP 5.1.1, WC 3.6.1, 256MB Memory Limit, PHP 7.3.4, Apache

Copy and paste the system status report from **WooCommerce > System Status** in WordPress admin.

`
### WordPress Environment ###

WordPress address (URL): https://thesandwichspotcafe.com
Site address (URL): https://thesandwichspotcafe.com
WC Version: 3.6.1
Log Directory Writable: ✔
WP Version: 5.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache
PHP Version: 7.3.4
PHP Post Max Size: 32 MB
PHP Time Limit: 60
PHP Max Input Vars: 1000
cURL Version: 7.19.7
NSS/3.27.1

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.0.38-MariaDB
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Your server does not have the SoapClient class enabled - some gateway plugins which use SOAP may not work as expected.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.6.1
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 59.81MB
Database Data Size: 49.84MB
Database Index Size: 9.97MB
wp_woocommerce_sessions: Data: 1.02MB + Index: 0.02MB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_order_items: Data: 0.25MB + Index: 0.11MB
wp_woocommerce_order_itemmeta: Data: 20.52MB + Index: 3.03MB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB
wp_cart_notices: Data: 0.02MB + Index: 0.00MB
wp_commentmeta: Data: 0.00MB + Index: 0.00MB
wp_comments: Data: 0.44MB + Index: 0.47MB
wp_gf_draft_submissions: Data: 0.02MB + Index: 0.02MB
wp_gf_entry: Data: 0.02MB + Index: 0.03MB
wp_gf_entry_meta: Data: 0.06MB + Index: 0.05MB
wp_gf_entry_notes: Data: 0.02MB + Index: 0.03MB
wp_gf_form: Data: 0.02MB + Index: 0.00MB
wp_gf_form_meta: Data: 0.03MB + Index: 0.00MB
wp_gf_form_revisions: Data: 0.02MB + Index: 0.03MB
wp_gf_form_view: Data: 0.22MB + Index: 0.17MB
wp_imagify_files: Data: 0.02MB + Index: 0.08MB
wp_imagify_folders: Data: 0.02MB + Index: 0.03MB
wp_jckwds: Data: 0.09MB + Index: 0.00MB
wp_links: Data: 0.00MB + Index: 0.00MB
wp_options: Data: 2.48MB + Index: 0.09MB
wp_postmeta: Data: 15.52MB + Index: 4.03MB
wp_posts: Data: 6.36MB + Index: 0.33MB
wp_rg_form: Data: 0.02MB + Index: 0.00MB
wp_rg_form_meta: Data: 0.03MB + Index: 0.00MB
wp_rg_form_view: Data: 0.20MB + Index: 0.17MB
wp_rg_incomplete_submissions: Data: 0.02MB + Index: 0.02MB
wp_rg_lead: Data: 0.02MB + Index: 0.03MB
wp_rg_lead_detail: Data: 0.06MB + Index: 0.06MB
wp_rg_lead_detail_long: Data: 0.02MB + Index: 0.00MB
wp_rg_lead_meta: Data: 0.02MB + Index: 0.05MB
wp_rg_lead_notes: Data: 0.02MB + Index: 0.03MB
wp_say_what_strings: Data: 0.02MB + Index: 0.00MB
wp_termmeta: Data: 0.02MB + Index: 0.03MB
wp_terms: Data: 0.00MB + Index: 0.01MB
wp_term_relationships: Data: 0.00MB + Index: 0.01MB
wp_term_taxonomy: Data: 0.00MB + Index: 0.00MB
wp_translations: Data: 0.28MB + Index: 0.09MB
wp_translations_log: Data: 0.02MB + Index: 0.02MB
wp_usermeta: Data: 1.52MB + Index: 0.36MB
wp_users: Data: 0.02MB + Index: 0.03MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB
wp_woocommerce_pickup_locations_geodata: Data: 0.02MB + Index: 0.05MB
wp_woos_search_trends: Data: 0.02MB + Index: 0.02MB
wp_yoast_seo_links: Data: 0.02MB + Index: 0.02MB
wp_yoast_seo_meta: Data: 0.06MB + Index: 0.00MB

### Post Type Counts ###

attachment: 142
nav_menu_item: 25
nectar_slider: 1
page: 9
post: 1
postman_sent_mail: 250
product: 49
product_variation: 24
scheduled-action: 28
shop_coupon: 2
shop_order: 861
shop_order_refund: 13
tm_global_cp: 5
wc_pickup_location: 2
wpseo_locations: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (37) ###

ManageWP - Worker: by GoDaddy – 4.7.7
Gravity Forms: by rocketgenius – 2.4.6.15
Additional Charge: by Yong Qui Zheng – 1.0 – Not tested with the active version of WooCommerce
Admin Menu Editor Pro: by Janis Elsts – 2.6.1
Akismet Anti-Spam: by Automattic – 4.1.1
Business Hours Indicator: by Maarten Belmans – 2.2.2
Classic Editor: by WordPress Contributors – 1.4
Webcraftic Disable Admin Notices Individually: by Webcraftic  – 1.0.6
WooCommerce Delivery Slots by Iconic: by Iconic – 1.7.15 – Not tested with the active version of WooCommerce
Imagify: by WP Media – 1.8.4.1
Invisible reCaptcha: by Mihai Chelaru – 1.2.3
Salient WPBakery Page Builder: by Michael M - WPBakery.com | Modified by ThemeNectar – 5.6
Loco Translate: by Tim Whitlock – 2.2.2
My Custom Functions: by Space X-Chimp – 4.38
Postman SMTP: by Jason Hendriks – 1.7.2
Really Simple SSL: by Rogier Lankhorst
Mark Wolters – 3.1.5

Tracking Code Manager: by IntellyWP – 1.11.9 – Not tested with the active version of WooCommerce
Transposh Translation Filter: by Team Transposh – 1.0.4.1
UpdraftPlus - Backup/Restore: by UpdraftPlus.Com
DavidAnderson – 2.16.5.24

WooCommerce Cart Add-Ons: by WooCommerce – 1.5.19 – Not tested with the active version of WooCommerce
WooCommerce Checkout Field Editor: by WooCommerce – 1.5.17 – Not tested with the active version of WooCommerce
WooCommerce Conditional Shipping and Payments: by SomewhereWarm – 1.5.2
WooCommerce Email Customizer: by WooCommerce – 1.1.9 – Not tested with the active version of WooCommerce
WooCommerce Stripe Gateway: by WooCommerce – 4.1.16
WooCommerce Google Analytics Pro: by SkyVerge – 1.7.1
WooCommerce Product Details Customiser: by jameskoster – 0.2.0 – Not tested with the active version of WooCommerce
Woocommerce Products Per Page: by Jeroen Sormani – 1.2.6 – Not tested with the active version of WooCommerce
WooCommerce Sequential Order Numbers Pro: by SkyVerge – 1.13.1
WooCommerce Sequential Order Numbers Pro: by SkyVerge – 1.13.1
WooCommerce Local Pickup Plus: by SkyVerge – 2.4.0 – Not tested with the active version of WooCommerce
WooCommerce TM Extra Product Options: by themeComplete – 4.9
WooCommerce URL Coupons: by SkyVerge – 2.7.3 – Not tested with the active version of WooCommerce
WooCommerce: by Automattic – 3.6.1
Yoast SEO: by Team Yoast – 11.0 – Not tested with the active version of WooCommerce
WP Rocket: by WP Media – 3.3.1.1
Yoast SEO: Local: by Team Yoast and Arjan Snaterse – 10.0.1
YITH WooCommerce Customize My Account Page: by YITH – 2.5.8

### Dropin Plugins (1) ###

advanced-cache.php: advanced-cache.php

### Settings ###

API Enabled: ✔
Force SSL: ✔
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: ✔

### WC Pages ###

Shop base: #65 - /shop/
Cart: #66 - /cart/
Checkout: #67 - /checkout/
My account: #68 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Salient
Version: 10.0.4
Author URL: https://themeforest.net/user/themenectar
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: /home/cocomifk/domains/thesandwichspotcafe.com/public_html/wp-content/plugins/woocommerce-tm-extra-product-options/templates/cart/cart-item-data.php
salient/woocommerce/checkout/form-checkout.php
salient/woocommerce/checkout/review-order.php
salient/woocommerce/checkout/terms.php
salient/woocommerce/content-product.php
salient/woocommerce/content-single-product.php
salient/woocommerce/loop/add-to-cart.php
salient/woocommerce/loop/loop-start.php
salient/woocommerce/myaccount/form-login.php
salient/woocommerce/single-product/product-image.php
salient/woocommerce/single-product/tabs/description.php
salient/woocommerce/single-product/tabs/tabs.php

### Local Pickup Plus ###

Status: ✔ Enabled
Geocoding: ✔
Pickup Locations: 2
Pickup Appointments: –
Pickup Location Tax: ✔
Checkout Display: Pickup selection mode: per orderCart item handling mode: automatic groupingDefault cart item handling: pickup items

### Action Scheduler ###

Complete: 27
Oldest: 2019-04-17 10:15:23 -0700
Newest: 2019-04-18 13:40:24 -0700

Pending: 1
Oldest: 2019-04-24 10:15:23 -0700
Newest: 2019-04-24 10:15:23 -0700

Canceled: 0
Oldest: –
Newest: –

In-progress: 0
Oldest: –
Newest: –

Failed: 0
Oldest: –
Newest: –

`
</details>

**Comment by rodrigoprimo** [comment]
Thanks for your report, @co50. I was not able to reproduce this issue on my local environment. I tested with different orders with different status. The refund field is not grayed out and I can successfully process the refund.

Are you able to reproduce this problem in a clean WooCommerce core environment? Could you please provide the links to the reports of other users experiencing the same problem? They might contain more details that might help.

**Comment by willgorham** [comment]
I believe this is intentional. See #23379, specifically https://github.com/woocommerce/woocommerce/issues/23379#issuecomment-484610279 for reasoning.

**Comment by co50** [comment]
I've heard from multiple colleagues who have Woo with the PayPal gateway who are running into this issue. My client has Stripe but same issue....I read thread 23379 - it appears that Woo changed the way refunds work and didn't tell anyone? If I'm I reading this right, now if a customer needs the whole order cancelled and refunded we have to go line by line and "refund" the items? How do you even do that? For as long as I've used Woo, qty and price adjustments are informational and don't auto-refund anything....that would be great if that worked now, but there still needs to be a way faster method of refunding an entire order that has multiple line items.

**Comment by co50** [comment]
Following up - and I'm an idiot for believing this would work, I first tried changing the line item qty to "0" which removed the line items, offering no option to refund, nor did it refund automatically. I then tried deleting the line item, same result. 

Mike Joley in the other thread instructed to "refund the line items" when this is not a function Woo offers. So we're back to square one. How do we refund Stripe charges using the order screen as we've been able to for the last 2 years up until now?

**Comment by willgorham** [comment]
I agree this isn't the *most* user-friendly way to require refunds be handled. But it does work. 

Click Refund. Then, in the line item boxes that appear, enter the number of items that you want to refund.

(*yes, without a label on the inputs this is unintuitive. I would expect this box to be the number of items you want to update the order to, e.g. 0 if fully refunded*). 

You should see the total refund box update and can then click Refund via Stripe.

**Comment by co50** [comment]
YES  - two comments here....

1) Agreed, this is definitely not intuitive and this change is big enough the Woo should have sent out an email or some highly visible notification in the admin after the update. It's going to confuse a lot of people. And there really *really* needs to be a full order refund option. For an order with 20+ items, it's insane to have to click REFUND and then change the qty to refund to "1" twenty times.

2) VERY NICE that this feature is finally added. Now if it could do the same with billing (for customers who saved their CC info), that would be cool.

Thanks for the explanation, Will.

**Comment by mikejolley** [comment]
As Will pointed out, this is intentional to prevent users breaking reports (which is common) due to not knowing what proportion of the refund applies to which items and which taxes. I've added an explanation section to https://github.com/woocommerce/woocommerce/wiki/3.6.x-notes-FAQ#the-refund-input-field-is-grayed-out-when-viewing-orders-and-trying-to-refund-items. If a full refund button is needed that is something to consider for 3.7 and should be logged separately.

Thanks

**Comment by co50** [comment]
I dunno, Mike...is undoing a regression now called a feature request? Geez, I hope this makes it into 3.6.1. So far every single client of mine with Woo has called me to ask how they are supposed to do refunds all of a sudden. It's a huge regression, made worse by the non-acknowledgement that it's a regression. I get why the change was made, but the way it's been done is awful.

**Comment by mikejolley** [comment]
@co50 Maybe you could ask them how they expect full refunds to work? If it could avoid the manual price field, that's what we'd need. Your clients reporting will not be accurate if manually inputting that refund amount..

If they don't enable taxes this is a non-issue. The field is not greyed.

**Comment by co50** [comment]
Yes, great question. I asked them, polled the Woo community, held several focus groups, watched 3 related TED talks, and this is what we came up with:

When REFUND is clicked, a "Refund" column appears in front of the thumbnail for each line item. Going down the column are checkboxes in front of each line item, with another checkbox at the top for "entire order". Clicking a line item with a qty higher than 1, displays a qty field under it where the number to refund can be entered - it is pre-populated with the qty purchased (eg. 4 bought, 4 appears in the field) and the admin can change the number or leave as is to refund the entire line item. Plenty of space to borrow from the Item column to fit this in and it's far more intuitive than the current model.

Again, just to reiterate, the issue isn't that you've decided to force accurate tax record keeping. Kudos to you for doing that. It's the manner in which it was done is not intuitive and NOBODY used to the old way is going to be able to figure this out on their own, so the lack of instruction was a swing & miss.

**Comment by mikejolley** [comment]
@co50 Thanks for the sarcasm :)

I'm more curious about how the field was used previously. You say they were doing full refunds but was this always the case? Were shipping cost for example refunded? Were they "auto" or "manual" refunds? Did they ever do partial refunds using that field instead of refunding by line item?

Pre-populating fields is an option here, with refund amount being filled defaulting to the full balance. That would however slow down a partial refund.

>  It's the manner in which it was done is not intuitive and NOBODY used to the old way is going to be able to figure this out on their own

Maybe we disagree here but I'd argue broken reporting is way more serious than a more complicated refund flow. Obviously the order screens are not ideal at present and need a revamp, but that's a large project whilst this was a quick fix to prevent users breaking reporting and having no way to correct it!

**Comment by co50** [comment]
You say sarcasm, but we really did all that stuff. I come with receipts, man. :)

Previously, my clients did the calculation of what was owed for not-entire-order-refunds themselves - always auto (Stripe, PP, and Auth.net AIM are gateways in use), and then they'd delete the line item from the order. Because Woo never did the calculation for them, they got used to doing it themselves. Now Woo 3.6 does/forces that for them, and we welcome that but a) the way it's done is strange, b) it's still not a locked down system. One can still add/delete items to the order without it forcing a charge/refund, so it's unusual for Woo to suddenly care so much about preventing auto refunds without knowing exactly what needs to be refunded, but there is no mechanism to auto-charge or instruct to do so if stuff is added. If this is all about accurate tax calculation, it still fails if an item is deleted, qty is edited, an item is added, etc. All that stuff is still informational and not tied to the 3rd party gateway.

So it confuses my clients into thinking it's idiot-proof when it's not. From a consistency point of view, if we're going to integrate with a 3rd party gateway, ideally Woo should enforce proper accounting all the time, not some of the time - ensuring that any changes to the order include a charge (or explicit instructions to do so if CC not saved) or refund whenever applicable and the option to override/ignore that is fine to include. But at least the client knows that Woo is going to tell them what needs to be refunded/charged every time.

I guess I didn't think the broken reporting was a big deal because it's been broken for the entire existence of the auto refund feature, right? But I felt the improper refund flow was urgent because my clients all rushed my phone about it....I didn't even know about the change until we started getting calls asking us to perform refunds ("Uhh, why can't you do this yourself?"). LOL

Ultimately, Mike, if nothing else changes, I recommend adopting a best practice where anytime you guys change functionality such as this, you add a quick pop-up tour (like you did I think when the order detail screen was changed several versions ago) that simply explains how to do whatever changed. Would have been super helpful on this issue.

**Comment by galapogos01** [comment]
How do I vote for this?  Very confusing change with a poor UX.

**Comment by agaleski** [comment]
> Maybe we disagree here but I'd argue broken reporting is way more serious than a more complicated refund flow.

@mikejolley 
Depends. In our case we never use the reporting anyways because it is broken anyways if you use many additional plugins. But refunds are happening VERY often due to the amount of orders we do per month (around 200.000 Orders/Month). The new method caused a lot of friction in our Customer-Care departments and raised significantly the amount of time spend on refunds which at the end costs us man hours = money.

From your wiki:
> Changing order status to 'refunded' which creates a refund on all line items in the order.

Yea, theoretically. But this does not trigger a refund with the payment provider in all payment gateways. Which forced us in the past already to disable the "change to status refunded manually" functionality as a whole to avoid human mistakes in our Customer-Care departments.

Overall I do understand what you want to achieve but it feels very much like a change that was made by someone who has never actually worked with the functionality in a real environment.

This should have been either discussed or anounced prior to the release!

**Comment by co50** [comment]
Agreed. Just a terrible implementation in every aspect. From not making a HUGE deal about a MAJOR change to how refunds are done via gateways, to the actual way it's done, which is a major regression. Disastrous roll-out.

BEFORE: Click 1) Refund, 2) Enter order total, 3) Click Refund via [gateway]

AFTER: Click 1) Refund, 2) Change the quantity to the # ordered for every single line item on the order, 3) Add in any manual 3rd party charges (gratuity, service fees, etc.), 4) Click Refund via [gateway]

How the heck is this better for Woo users? Takes longer, introduces more room for human error, counter-intuitive for newbies (eg. is it the qty I want refunded or the qty they have after the refund (0)?).

**Comment by agaleski** [comment]
https://raw.githubusercontent.com/woocommerce/woocommerce-gateway-stripe/master/changelog.txt

Changelog 4.2.2 – 2019-06-26
* Fix – Changing an order status to “Cancelled” or “Refunded” **will no longer refund the payment**, will only void the payment if it was just authorized.

@mikejolley I just wanted to pin this here because your wiki entry ("Changing order status to 'refunded' which creates a refund on all line items in the order.") is proven factually wrong by **STRIPE** itself.

**Comment by addedlovely** [comment]
@mikejolley Just to add - getting a few confused clients on this implementation. Is simple enough when explained - but most frequent use case for them is refunding an entire order not individual line items. 

Perhaps there's are two buttons. 'Partial Refund' and 'Full Refund' - the latter pre-populates the fields, or just processes the refund.
